proposals/254-padding-negotiation.txt

```diff
@@ -436,42 +177,113 @@ increasing bin width allows the histograms to most accurately represent
 small interpacket delay (where accuracy is needed), and devote less
 accuracy to larger timescales (where accuracy is not as important).
 
-3.2.4. Token removal and refill
+3.1.2 Histogram Token Removal
+
+Tokens can be optionally removed from histogram bins whenever a padding or
+non-padding packet is sent. With this token removal, the histogram functions
+as an overall target delay distribution for the machine while it is in that
+state.
+
+If token removal is enabled, when a padding packet is sent, a token is removed
+from the bin corresponding to the target delay. When a non-padding packet is
+sent, the actual delay from the previous packet is calculated, and the
+histogram bin corresponding to that delay is inspected.  If that bin has
+tokens remaining, it is decremented.
+
+If the bin has no tokens left, the state removes a token from a different bin,
+as specified in its token removal rule. The following token removal options
+are defined:
+  * None -- Never remove any tokens
+  * Exact -- Only remove from the target bin, if it is empty, ignore it.
+  * Higher -- Remove from the next higher non-empty bin
+  * Lower -- Remove from the next higher non-empty bin
+  * Closest -- Remove from the closest non-empty bin by index
+  * Closest_time -- Remove from the closest non-empty bin by index, by time
+
+When all bins are empty in a histogram, the padding machine emits the internal
+"bins empty" event to itself.
+
+3.2. State Machine Selection
+
+Clients will select which of the defined available padding machines to use
+based on the conditions that these machines specify. These conditions include:
+  * How many hops the circuit must be in order for the machine to apply
+  * If the machine requires vanguards to be enabled to apply
+  * The state the circuit must be in for machines to apply (building,
+    relay early cells remaining, opened, streams currently attached).
+  * If the circuit purpose matches a set of purposes for the machine.
+  * If the target hop of the machine supports circuit padding.
+
+Clients will only select machines whose conditions fully match given circuits.
+
+A machine is represented by a positive number that can be thought of as a "menu
+option" through the list of padding machines. The currently supported padding
+state machines are:
+
+        [1]: CIRCPAD_MACHINE_CIRC_SETUP
```

Comment:
Should we specify more in-depth what this state machine does, or how it works?

Reply:
Not yet. It should work better first, I think.